// react-fluid-gallery/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _raf = _interopRequireDefault(require("raf"));

var _reactSizeme = _interopRequireDefault(require("react-sizeme"));

var _reactSwipeable = _interopRequireDefault(require("react-swipeable"));

var _fluidGallery = _interopRequireDefault(require("./fluid-gallery"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var noop = function noop() {
  return undefined;
};

var ReactFluidGallery =
/*#__PURE__*/
function (_Component) {
  _inherits(ReactFluidGallery, _Component);

  function ReactFluidGallery() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, ReactFluidGallery);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(ReactFluidGallery)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_current", typeof _this.props.startAt !== 'undefined' ? _this.props.startAt : Math.random() * _this.props.slides.length | 0);

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_containerRef", function (ref) {
      _this._container = ref;
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_canvasRef", function (ref) {
      _this._canvas = ref;
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_onWheel", function (event) {
      event.preventDefault();

      _this._gallery.onScroll(event);
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_onSwiped", function (event, deltaX, deltaY, isFlick) {
      if (isFlick) {
        deltaY *= 5;
      }

      _this._gallery.onScroll(_objectSpread({}, event, {
        deltaY: deltaY
      }));
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_onResize", function () {
      _this._canvas.width = _this._container.clientWidth;
      _this._canvas.height = _this._container.clientHeight;

      if (_this._gallery) {
        _this._gallery.resize();
      }
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_tick", function () {
      var onChange = _this.props.onChange;

      _this._gallery.update();

      _this._gallery.render();

      if (_this._current !== _this._gallery.currentSlideIndex) {
        _this._current = _this._gallery.currentSlideIndex;
        onChange(_this._current);
      }

      _this._tickRaf = (0, _raf.default)(_this._tick);
    });

    return _this;
  }

  _createClass(ReactFluidGallery, [{
    key: "componentWillReceiveProps",
    value: function componentWillReceiveProps(props) {
      this._onResize();
    }
  }, {
    key: "componentDidMount",
    value: function componentDidMount() {
      window.addEventListener('resize', this._onResize);

      this._reset(this.props);

      this._tick();
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      window.removeEventListener('resize', this._onResize);

      if (this._tickRaf) {
        _raf.default.cancel(this._tickRaf);

        this._tickRaf = null;
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          slides = _this$props.slides,
          startAt = _this$props.startAt,
          onChange = _this$props.onChange,
          style = _this$props.style,
          rest = _objectWithoutProperties(_this$props, ["slides", "startAt", "onChange", "style"]);

      return _react.default.createElement(_reactSwipeable.default, _extends({
        style: _objectSpread({
          width: '100%',
          height: '100%',
          overflow: 'hidden'
        }, style)
      }, rest, {
        stopPropagation: true,
        preventDefaultTouchmoveEvent: true,
        onSwiped: this._onSwiped,
        innerRef: this._containerRef
      }), _react.default.createElement("canvas", {
        ref: this._canvasRef,
        onWheel: this._onWheel,
        style: {
          width: '100%',
          height: '100%'
        }
      }));
    }
  }, {
    key: "_reset",
    value: function _reset(props) {
      var slides = props.slides;

      this._onResize();

      this._gallery = new _fluidGallery.default({
        canvas: this._canvas,
        slides: slides,
        current: this._current
      });
    }
  }]);

  return ReactFluidGallery;
}(_react.Component);

_defineProperty(ReactFluidGallery, "propTypes", {
  slides: _propTypes.default.arrayOf(_propTypes.default.string).isRequired,
  startAt: _propTypes.default.number,
  onChange: _propTypes.default.func,
  style: _propTypes.default.object,
  size: _propTypes.default.shape({
    width: _propTypes.default.number,
    height: _propTypes.default.number
  })
});

_defineProperty(ReactFluidGallery, "defaultProps", {
  onChange: noop,
  style: {}
});

var _default = (0, _reactSizeme.default)({
  monitorWidth: true,
  monitorHeight: true
})(ReactFluidGallery);

exports.default = _default;
